core/src/test/java/org/infinispan/stream/DistributedWriteBehindStreamIteratorTest.java

```diff
@@ -0,0 +1,129 @@
+
+            new DistributedWriteBehindStreamIteratorTest().async(true).shared(true).cacheMode(CacheMode.DIST_SYNC),
+            new DistributedWriteBehindStreamIteratorTest().async(false).shared(true).cacheMode(CacheMode.DIST_SYNC),
+            new DistributedWriteBehindStreamIteratorTest().async(true).shared(false).cacheMode(CacheMode.DIST_SYNC),
+            new DistributedWriteBehindStreamIteratorTest().async(false).shared(false).cacheMode(CacheMode.DIST_SYNC),
+      };
+   }
+
+   @Override
+   protected Object[] parameterValues() {
+      return concat(super.parameterValues(), asyncStore, sharedStore);
+   }
+
+   @Override
+   protected String[] parameterNames() {
+      return concat(super.parameterNames(), "asyncStore", "sharedStore");
+   }
+
+   @Override
+   protected void enhanceConfiguration(ConfigurationBuilder builder) {
+      DummyInMemoryStoreConfigurationBuilder dimscb = builder.persistence().addStore(DummyInMemoryStoreConfigurationBuilder.class);
+      if (sharedStore) {
+         dimscb.shared(true);
+      }
+      if (asyncStore) {
+         dimscb.storeName(getTestName())
+               .async().enable();
+      }
+   }
+
+   @Test
+   public void testStayLocalIfAllSegmentsPresentLocallyWithWriteBehindStoreWithReHash() throws Exception {
+      testStayLocalIfAllSegmentsPresentLocallyWithWriteBehindStore(true);
+   }
+
+   @Test
+   public void testStayLocalIfAllSegmentsPresentLocallyWithWriteBehindStoreWithoutRehash() throws Exception {
+      testStayLocalIfAllSegmentsPresentLocallyWithWriteBehindStore(false);
+   }
+
+   protected void testStayLocalIfAllSegmentsPresentLocallyWithWriteBehindStore(boolean rehashAware) {
+      Cache<Object, String> cache1 = cache(1, CACHE_NAME);
+
+      ClusterStreamManager clusterStreamManager = replaceWithSpy(cache1);
+
+      for (Cache<Object, String> cache : this.<Object, String>caches(CACHE_NAME)) {
+         MagicKey key = new MagicKey(cache);
+         cache.put(key, key.toString());
+      }
+
+      // Remember that segment ownership is as {0, 1}, {1, 2}, {2, 1}
+
+      CacheStream<Map.Entry<Object, String>> stream = cache1.entrySet().stream();
+      if (!rehashAware) stream = stream.disableRehashAware();
+
+      int invocationCount;
+      if (cacheMode.isReplicated()) {
+         Map<Object, String> entries = mapFromIterator(stream.iterator());
+         // We should not see keys from other segments, but there may be segments without any keys
+         assertEquals(caches(CACHE_NAME).size(), entries.size());
+         invocationCount = cacheManagers.size() - 1;
+      } else {
+         // Distributed cache
+         // Cache1 owns 1 (primary) and 2 (backup)
+         // When it is a write behind shared store it will have to go remote otherwise will stay local
+         Map<Object, String> entries = mapFromIterator(stream.filterKeySegments(IntSets.immutableSet(2)).iterator());
```

Comment:
Question: is the optimization performed only when all the requested segments are local, or is it also performed for backup-owned segments when the application requests all the segments? If it's the former, then why? If it's the latter, then I think it would be best to stream all segments in dist caches as well.

Reply:
If the local node owns all requested segments it will stay local for them - unless shared write behind mode. Are you asking about a subset? In this case it only does primary on the local node. The new publisher however does backups even when it doesn't own all of them.

Reply:
Yes, I was asking about dist caches, where not all requested segments are local. Glad to hear that the new publisher optimizes backups owned by the originator in that case as well.